src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
+                {
+                    if (cBoss->IsAlive())
+                        return CheckIsMovementFinished(); // in case a horseman dies while moving (unlikely but possible especially in non-335 branch)
+                }
+                else
+                {
+                    TC_LOG_WARN("scripts", "FourHorsemenAI: %s died but horseman with id %d is not present", me->GetName().c_str(), uint32(boss));
+                    ResetEncounter();
                 }
 
-                if (checkAllDead)
-                    return !Thane->IsAlive() && !Lady->IsAlive() && !Baron->IsAlive() && !Sir->IsAlive();
-            }
-            return false;
+            EncounterCleared();
         }
 
-        void BeginFourHorsemenMovement()
+        void MovementInform(uint32 type, uint32 i) override
         {
-            movementStarted = true;
-            me->SetReactState(REACT_PASSIVE);
-            me->SetWalk(false);
-            me->SetSpeed(MOVE_RUN, me->GetSpeedRate(MOVE_RUN), true);
-
-            switch (id)
+            if (type != POINT_MOTION_TYPE)
+                return;
+            if (i < 3)
+                _nextMovement = i; // delay to next updateai to prevent it from instantly expiring
+            else
             {
-                case HORSEMEN_THANE:
-                    me->GetMotionMaster()->MovePoint(0, WaypointPositions[0]);
-                    break;
-                case HORSEMEN_LADY:
-                    me->GetMotionMaster()->MovePoint(3, WaypointPositions[3]);
-                    break;
-                case HORSEMEN_BARON:
-                    me->GetMotionMaster()->MovePoint(6, WaypointPositions[6]);
-                    break;
-                case HORSEMEN_SIR:
-                    me->GetMotionMaster()->MovePoint(9, WaypointPositions[9]);
-                    break;
+                _myMovementFinished = true;
+                CheckIsMovementFinished();
             }
         }
 
-        void MovementInform(uint32 type, uint32 point) override
+        void UpdateAI(uint32 diff) override
         {
-            if (type != POINT_MOTION_TYPE)
-                return;
+            if (_nextMovement)
+            {
+                me->GetMotionMaster()->MovePoint(_nextMovement + 1, _initialPath[_nextMovement], true);
+                _nextMovement = 0;
+            }
+            _UpdateAI(diff);
+        }
 
-            if (point == 2 || point == 5 || point == 8 || point == 11)
+        virtual void BeginFighting() = 0;
+        virtual void _UpdateAI(uint32 /*diff*/) = 0;
+
+    private:
+        const Horseman _which;
+        const Position* _initialPath;
+        bool _myMovementFinished;
+        uint8 _nextMovement;
+        uint32 _timeDied;
+    protected:
+        bool _ourMovementFinished;
+};
+
+static const Position baronPath[3] = { { 2552.427f, -2969.737f, 241.3021f }, { 2566.759f, -2972.535f, 241.3217f }, { 2584.32f, -2971.96f, 241.3489f } };
```

Comment:
it might be better to move the paths on top of the file with all the other definitions